src/librustdoc/formats/cache.rs

```diff
@@ -164,10 +164,12 @@ impl Cache {
             };
             let name = e.name(tcx);
             let extern_url = extern_html_root_urls.get(&*name.as_str()).map(|u| &**u);
-            self.extern_locations
-                .insert(n, (name, src_root, extern_location(e, extern_url, &dst, tcx)));
-
             let did = DefId { krate: n, index: CRATE_DEF_INDEX };
+            self.extern_locations.insert(
+                n,
+                (name, src_root, extern_location(e, extern_url, tcx.get_attrs(did), &dst, tcx)),
+            );
```

Comment:
Doesn't have to be here, but it would be nice to store the CrateNum here instead of the name and attributes. Both of those are available from the tcx.

Actually all of this looks like it could be calculated on-demand with only the CrateNum, tcx, and extern_html_root_urls.

Reply:
I'm not sure I understood correctly:
Yo would like to replace `crate external_paths: FxHashMap<DefId, (Vec<String>, ItemType)>` by `crate extern_html_urls: FxHashMap<DefId, &BTreeMap<String, String>>`, get `name` and `attributes` from `tcx`, combine them with `cache.extern_html_root_url[CrateNum]` and then compute `extern_locations` with all this ?

Reply:
I would like to get rid of `extern_locations` altogether, and calculate `name`, `src_root`, and `external_location` on demand. `external_paths` can stay the same.

I'll open an issue for it, it doesn't belong here.

Reply:
Oipened https://github.com/rust-lang/rust/issues/84588.